File: 6_Simple_Algebraic_Data_Types/shape.cpp
#include "shape.h"

Circle::Circle(float r) : radius(r) {}

float Circle::area() const {
    return M_PI * radius * radius;
}

float Circle::circ() const {
    return 2.0 * M_PI * radius;
}

Rect::Rect(float w, float h) : width(w), height(h) {}

float Rect::area() const {
    return width * height;
}

float Rect::circ() const {
    return 2.0 * (width + height);
}

Square::Square(float s) : side(s) {}

float Square::area() const {
    return side * side;
}

float Square::circ() const {
    return 4.0 * side;
}

File: 6_Simple_Algebraic_Data_Types/shape.hs

data Shape = Circle Float | Rect Float Float | Square Float 
    deriving (Show)

area :: Shape -> Float
area (Circle r) = pi * r * r
area (Rect d h) = d * h
area (Square s) = s * s

circ :: Shape -> Float
circ (Circle r) = 2 * pi * r
circ (Rect d h) = 2 * (d + h)
circ (Square s) = 4 * s

main = do
    let shapes = [Circle 5.0, Rect 4.0 6.0, Square 5.0]
    mapM_ (\shape -> print $ "Area: " ++ show (area shape)
                     ++ ", Circumference: " ++ show (circ shape)) shapes


